<a href="https://colab.research.google.com/github/sudhans18/ABDA/blob/main/04_Linguistic_Complexity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import os
import pickle
import json
import warnings

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

warnings.filterwarnings("ignore")

# Reproducibility
np.random.seed(42)

# Project paths
BASE_DIR = "/content/drive/MyDrive/cmu_mosei"

LABSE_READY_PATH = os.path.join(
    BASE_DIR,
    "CMU_MOSEI_LaBSE_ready_v2.pkl"
)

print("Notebook 4: Linguistic Complexity")
print("=" * 55)
print("Base directory:", BASE_DIR)
print("Input file:", LABSE_READY_PATH)
print("File exists:", os.path.exists(LABSE_READY_PATH))

Notebook 4: Linguistic Complexity
Base directory: /content/drive/MyDrive/cmu_mosei
Input file: /content/drive/MyDrive/cmu_mosei/CMU_MOSEI_LaBSE_ready_v2.pkl
File exists: True


Load the prepared CMU-MOSEI dataset

In [ ]:
with open(LABSE_READY_PATH, "rb") as f:
    data = pickle.load(f)

print("Loaded object type:", type(data))

if isinstance(data, pd.DataFrame):
    df = data.copy()
    print("Dataset loaded directly as DataFrame.")
else:
    print("Dataset is not a DataFrame.")
    print("Object information:")
    print(data if not isinstance(data, dict) else data.keys())

Loaded object type: <class 'pandas.core.frame.DataFrame'>
Dataset loaded directly as DataFrame.


In [ ]:
print("Shape:", df.shape)
print("\nColumns:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nFirst 5 rows:")
display(df.head())

print("\nMissing values:")
print(df.isnull().sum())

print("\nText examples:")
for i, text in enumerate(df["text"].head(10)):
    print(f"{i}: {repr(text)}")

Shape: (23253, 13)

Columns:
['segment_id', 'label_idx', 'start', 'end', 'text', 'token_count', 'sentiment', 'happiness', 'sadness', 'anger', 'surprise', 'disgust', 'fear']

Data types:
segment_id      object
label_idx        int64
start          float64
end            float64
text            object
token_count      int64
sentiment      float64
happiness      float64
sadness        float64
anger          float64
surprise       float64
disgust        float64
fear           float64
dtype: object

First 5 rows:


,segment_id,label_idx,start,end,text,token_count,sentiment,happiness,sadness,anger,surprise,disgust,fear
0,--qXJuDtHPw,0,23.199,30.325,writer i see that a writer is somebody who has...,21,1.000000,0.666667,0.000000,0.0,0.0,0.0,0.000000
1,-3g5yACwYnA,0,82.753,100.555,key is part of the people that we use to solve...,49,1.000000,0.666667,0.666667,0.0,0.0,0.0,0.666667
2,-3g5yACwYnA,1,119.919,125.299,businesses that we do they've been able to fin...,23,0.666667,0.000000,0.000000,0.0,0.0,0.0,0.000000
3,-3g5yACwYnA,2,4.840,14.052,operations key polymer brings a technical aspe...,22,0.000000,0.666667,0.666667,0.0,0.0,0.0,0.333333
4,-3g5yACwYnA,3,13.211,27.521,internally we're a huge user of adhesives for ...,41,0.000000,0.333333,0.333333,0.0,0.0,0.0,0.000000



Missing values:
segment_id     0
label_idx      0
start          0
end            0
text           0
token_count    0
sentiment      0
happiness      0
sadness        0
anger          0
surprise       0
disgust        0
fear           0
dtype: int64

Text examples:
0: 'writer i see that a writer is somebody who has an incredible command of mechanics of the english language'
1: "key is part of the people that we use to solve those issues whether it's stretch or outdoor resistance or abrasions or different technical aspects that we really need to solve to get into new markets they've been able to bring solutions"
2: "businesses that we do they've been able to find solutions or at least bring some answers to the table"
3: "operations key polymer brings a technical aspect to our operation that we don't have internally we're"
4: "internally we're a huge user of adhesives for our operation called flocking and we don't have the technical inside compounding or the technical expertise to do th

## Set up linguistic complexity analysis

In [ ]:
text_series = df["text"].astype(str)

df["word_count_lc"] = text_series.str.split().str.len()
df["char_count_lc"] = text_series.str.len()

print("Word count statistics:")
print(df["word_count_lc"].describe())

print("\nCharacter count statistics:")
print(df["char_count_lc"].describe())

print("\nSelected percentiles — word count:")
print(df["word_count_lc"].quantile([0.01, 0.05, 0.25, 0.50, 0.75, 0.95, 0.99]))

print("\nZero-word samples:", (df["word_count_lc"] == 0).sum())

Word count statistics:
count    23253.000000
mean        21.122952
std         12.474859
min          1.000000
25%         13.000000
50%         18.000000
75%         26.000000
max        194.000000
Name: word_count_lc, dtype: float64

Character count statistics:
count    23253.000000
mean       112.300907
std         66.919655
min          2.000000
25%         68.000000
50%         98.000000
75%        140.000000
max        957.000000
Name: char_count_lc, dtype: float64

Selected percentiles — word count:
0.01     4.0
0.05     7.0
0.25    13.0
0.50    18.0
0.75    26.0
0.95    45.0
0.99    64.0
Name: word_count_lc, dtype: float64

Zero-word samples: 0


In [ ]:
print("Shortest samples:\n")

short_samples = df.nsmallest(10, "word_count_lc")[[
    "segment_id", "label_idx", "text", "word_count_lc"
]]

display(short_samples)

print("\nLongest samples:\n")

long_samples = df.nlargest(10, "word_count_lc")[[
    "segment_id", "label_idx", "text", "word_count_lc"
]]

display(long_samples)

Shortest samples:



,segment_id,label_idx,text,word_count_lc
1258,126872,2,seriously,1
1619,154449,0,hi,1
1864,194299,0,hi,1
2203,202431,0,hi,1
2816,221740,0,hi,1
3199,229967,0,hi,1
3271,233389,0,hi,1
3467,238023,0,hi,1
3803,243646,7,alright,1
3806,243646,10,case,1



Longest samples:



,segment_id,label_idx,text,word_count_lc
8814,83310,2,umm he wakes up twenty years later and uhh he ...,194
17309,dQ56b0bqmc8,2,at and all you need to do is print this form o...,194
8085,6UV6ktwbLoo,0,this is a quick video to show you how to do th...,188
7988,69707,3,cheesy it's just one of those really cheesy mo...,187
7894,6-0bcijTR8k,0,hi there today i have a very special announcem...,175
6357,30162,11,umm it's not up to par with anything on today'...,156
17165,d-Uw_uZyUys,2,but you know what it's not even just about him...,156
3798,243646,2,but umm what i wanted to say about this movie ...,149
14248,SqAiJrvHXNA,0,hi this is donnie welcome back to flooring my ...,138
1390,130149,5,release date anyways i don't know uhh pretty m...,135


In [ ]:
import spacy

MODEL_NAME = "en_core_web_sm"

try:
    nlp = spacy.load(MODEL_NAME)
    print(f"Loaded spaCy model: {MODEL_NAME}")
except OSError:
    print(f"{MODEL_NAME} not found. Installing...")
    !python -m spacy download en_core_web_sm
    nlp = spacy.load(MODEL_NAME)
    print(f"Loaded spaCy model: {MODEL_NAME}")

Loaded spaCy model: en_core_web_sm


In [ ]:
test_text = df["text"].iloc[0]

doc = nlp(test_text)

print("Text:")
print(test_text)

print("\nParsed tokens:")
for token in doc:
    print(
        f"{token.text:15} "
        f"POS={token.pos_:8} "
        f"DEP={token.dep_:12} "
        f"HEAD={token.head.text}"
    )

Text:
writer i see that a writer is somebody who has an incredible command of mechanics of the english language

Parsed tokens:
writer          POS=NOUN     DEP=npadvmod     HEAD=see
i               POS=PRON     DEP=nsubj        HEAD=see
see             POS=VERB     DEP=ROOT         HEAD=see
that            POS=SCONJ    DEP=mark         HEAD=is
a               POS=DET      DEP=det          HEAD=writer
writer          POS=NOUN     DEP=nsubj        HEAD=is
is              POS=AUX      DEP=ccomp        HEAD=see
somebody        POS=PRON     DEP=attr         HEAD=is
who             POS=PRON     DEP=nsubj        HEAD=has
has             POS=VERB     DEP=relcl        HEAD=somebody
an              POS=DET      DEP=det          HEAD=command
incredible      POS=ADJ      DEP=amod         HEAD=command
command         POS=NOUN     DEP=dobj         HEAD=has
of              POS=ADP      DEP=prep         HEAD=command
mechanics       POS=NOUN     DEP=pobj         HEAD=of
of              POS=ADP      DE

### Surface complexity features:
- word count
- character count
- average word length



In [ ]:
def extract_surface_features(text):
    """
    Extract basic surface-level linguistic measurements.

    These are descriptive features and are not interpreted
    independently as evidence of psychological state.
    """

    text = str(text).strip()

    if not text:
        return {
            "word_count": 0,
            "char_count": 0,
            "avg_word_length": 0.0
        }

    words = text.split()

    word_count = len(words)
    char_count = len(text)

    avg_word_length = (
        np.mean([len(word) for word in words])
        if words else 0.0
    )

    return {
        "word_count": word_count,
        "char_count": char_count,
        "avg_word_length": avg_word_length
    }

In [ ]:
for text in df["text"].head(5):
    features = extract_surface_features(text)

    print("\nText:", text)
    print(features)


Text: writer i see that a writer is somebody who has an incredible command of mechanics of the english language
{'word_count': 19, 'char_count': 105, 'avg_word_length': np.float64(4.578947368421052)}

Text: key is part of the people that we use to solve those issues whether it's stretch or outdoor resistance or abrasions or different technical aspects that we really need to solve to get into new markets they've been able to bring solutions
{'word_count': 42, 'char_count': 236, 'avg_word_length': np.float64(4.642857142857143)}

Text: businesses that we do they've been able to find solutions or at least bring some answers to the table
{'word_count': 19, 'char_count': 101, 'avg_word_length': np.float64(4.368421052631579)}

Text: operations key polymer brings a technical aspect to our operation that we don't have internally we're
{'word_count': 16, 'char_count': 101, 'avg_word_length': np.float64(5.375)}

Text: internally we're a huge user of adhesives for our operation called flocking an

### Dependency-based linguistic complexity
- dependency count
- dependency tree depth
- average dependency distance

#### What is dependency?
(eg) "The student solved the problem quickly."

A dependency parser tries to represent the grammatical relationships between words.

             solved
            /      \
        student    problem
          |           |
         The        the
                       \
                      quickly

student  → subject of → solved

problem  → object of  → solved

quickly  → modifier of → solved

In [ ]:
def extract_dependency_features(doc):
    """
    Extract syntactic complexity measures from a spaCy Doc.

    Features:
    - dependency_count:
        Number of syntactic dependency relations.
    - dependency_tree_depth:
        Maximum depth of the dependency structure.
    - avg_dependency_distance:
        Average linear distance between a token and its head.
    """

    tokens = [
        token for token in doc
        if not token.is_space and not token.is_punct
    ]

    if len(tokens) <= 1:
        return {
            "dependency_count": 0,
            "dependency_tree_depth": 0,
            "avg_dependency_distance": 0.0
        }

    # Number of dependency relations
    dependency_count = sum(
        1 for token in tokens
        if token.head != token
    )

    # --------------------------------------------------------
    # Dependency tree depth
    # --------------------------------------------------------

    def token_depth(token):
        depth = 0
        current = token

        visited = set()

        while current.head != current:
            if current.i in visited:
                break

            visited.add(current.i)
            depth += 1
            current = current.head

        return depth

    depths = [token_depth(token) for token in tokens]

    dependency_tree_depth = max(depths) if depths else 0

    # --------------------------------------------------------
    # Average dependency distance
    # --------------------------------------------------------

    distances = [
        abs(token.i - token.head.i)
        for token in tokens
        if token.head != token
    ]

    avg_dependency_distance = (
        float(np.mean(distances))
        if distances else 0.0
    )

    return {
        "dependency_count": dependency_count,
        "dependency_tree_depth": dependency_tree_depth,
        "avg_dependency_distance": avg_dependency_distance
    }

In [ ]:
for text in df["text"].head(5):

    doc = nlp(text)

    features = extract_dependency_features(doc)

    print("\nText:")
    print(text)

    print("Features:")
    print(features)


Text:
writer i see that a writer is somebody who has an incredible command of mechanics of the english language
Features:
{'dependency_count': 18, 'dependency_tree_depth': 9, 'avg_dependency_distance': 1.7222222222222223}

Text:
key is part of the people that we use to solve those issues whether it's stretch or outdoor resistance or abrasions or different technical aspects that we really need to solve to get into new markets they've been able to bring solutions
Features:
{'dependency_count': 43, 'dependency_tree_depth': 8, 'avg_dependency_distance': 2.4186046511627906}

Text:
businesses that we do they've been able to find solutions or at least bring some answers to the table
Features:
{'dependency_count': 19, 'dependency_tree_depth': 7, 'avg_dependency_distance': 2.0}

Text:
operations key polymer brings a technical aspect to our operation that we don't have internally we're
Features:
{'dependency_count': 17, 'dependency_tree_depth': 4, 'avg_dependency_distance': 3.235294117647059}



In [ ]:
texts = df["text"].astype(str).tolist()

surface_features = []
dependency_features = []

print("Processing", len(texts), "text segments...")

for i, (text, doc) in enumerate(
    zip(
        texts,
        nlp.pipe(texts, batch_size=64)
    )
):

    # Surface features
    surface = extract_surface_features(text)

    # Dependency features
    dependency = extract_dependency_features(doc)

    surface_features.append(surface)
    dependency_features.append(dependency)

    if (i + 1) % 2000 == 0:
        print(f"Processed {i + 1}/{len(texts)}")

print("Processing complete.")

Processing 23253 text segments...
Processed 2000/23253
Processed 4000/23253
Processed 6000/23253
Processed 8000/23253
Processed 10000/23253
Processed 12000/23253
Processed 14000/23253
Processed 16000/23253
Processed 18000/23253
Processed 20000/23253
Processed 22000/23253
Processing complete.


In [ ]:
surface_df = pd.DataFrame(surface_features)
dependency_df = pd.DataFrame(dependency_features)

linguistic_features = pd.concat(
    [surface_df, dependency_df],
    axis=1
)

print("Feature table shape:", linguistic_features.shape)

print("\nFeatures:")
print(linguistic_features.columns.tolist())

print("\nFirst 5 rows:")
display(linguistic_features.head())

Feature table shape: (23253, 6)

Features:
['word_count', 'char_count', 'avg_word_length', 'dependency_count', 'dependency_tree_depth', 'avg_dependency_distance']

First 5 rows:


,word_count,char_count,avg_word_length,dependency_count,dependency_tree_depth,avg_dependency_distance
0,19,105,4.578947,18,9,1.722222
1,42,236,4.642857,43,8,2.418605
2,19,101,4.368421,19,7,2.000000
3,16,101,5.375000,17,4,3.235294
4,32,192,5.031250,32,4,2.781250


In [ ]:
assert len(linguistic_features) == len(df), (
    f"Row mismatch: {len(linguistic_features)} vs {len(df)}"
)

print("Row count matches:", len(df))

# Attach segment ID temporarily for verification
verification_df = pd.DataFrame({
    "segment_id": df["segment_id"].values,
    "label_idx": df["label_idx"].values,
    "text": df["text"].values
})

verification_df = pd.concat(
    [verification_df, linguistic_features],
    axis=1
)

display(verification_df.head(10))

Row count matches: 23253


,segment_id,label_idx,text,word_count,char_count,avg_word_length,dependency_count,dependency_tree_depth,avg_dependency_distance
0,--qXJuDtHPw,0,writer i see that a writer is somebody who has...,19,105,4.578947,18,9,1.722222
1,-3g5yACwYnA,0,key is part of the people that we use to solve...,42,236,4.642857,43,8,2.418605
2,-3g5yACwYnA,1,businesses that we do they've been able to fin...,19,101,4.368421,19,7,2.000000
3,-3g5yACwYnA,2,operations key polymer brings a technical aspe...,16,101,5.375000,17,4,3.235294
4,-3g5yACwYnA,3,internally we're a huge user of adhesives for ...,32,192,5.031250,32,4,2.781250
5,-3g5yACwYnA,4,of things key brings those types of aspects to...,36,193,4.388889,35,10,2.371429
6,-3g5yACwYnA,5,solutions we have many new opportunities throu...,24,146,5.125000,23,9,2.260870
7,-3nNcZdcdvU,0,introduction speaker dr erma jean sims sonoma ...,32,175,4.500000,31,12,1.645161
8,-3nNcZdcdvU,1,it's important for teachers to understand thei...,16,116,6.312500,16,5,2.562500
9,-3nNcZdcdvU,2,noncompliance can result in legal liability to...,18,103,4.777778,17,6,2.117647


## Statistical validation

In [ ]:
print("Descriptive statistics:\n")

display(
    linguistic_features.describe().T
)

Descriptive statistics:



,count,mean,std,min,25%,50%,75%,max
word_count,23253.0,21.122952,12.474859,1.000000,13.000000,18.000000,26.000000,194.000000
char_count,23253.0,112.300907,66.919655,2.000000,68.000000,98.000000,140.000000,957.000000
avg_word_length,23253.0,4.375601,0.660444,1.666667,3.923077,4.307692,4.750000,9.750000
dependency_count,23253.0,20.595235,12.667285,0.000000,12.000000,18.000000,26.000000,200.000000
dependency_tree_depth,23253.0,5.528491,2.452002,0.000000,4.000000,5.000000,7.000000,25.000000
avg_dependency_distance,23253.0,2.462472,0.638813,0.000000,2.000000,2.400000,2.823529,8.320513


In [ ]:
print("Missing values:")
print(linguistic_features.isnull().sum())

print("\nInfinite values:")
print(np.isinf(linguistic_features.select_dtypes(include=np.number)).sum())

print("\nNegative values:")
print((linguistic_features < 0).sum())

Missing values:
word_count                 0
char_count                 0
avg_word_length            0
dependency_count           0
dependency_tree_depth      0
avg_dependency_distance    0
dtype: int64

Infinite values:
word_count                 0
char_count                 0
avg_word_length            0
dependency_count           0
dependency_tree_depth      0
avg_dependency_distance    0
dtype: int64

Negative values:
word_count                 0
char_count                 0
avg_word_length            0
dependency_count           0
dependency_tree_depth      0
avg_dependency_distance    0
dtype: int64


In [ ]:
correlation_matrix = linguistic_features.corr()

print("Feature correlation matrix:")
display(correlation_matrix.round(3))

Feature correlation matrix:


,word_count,char_count,avg_word_length,dependency_count,dependency_tree_depth,avg_dependency_distance
word_count,1.000,0.981,-0.030,0.997,0.674,0.531
char_count,0.981,1.000,0.137,0.978,0.696,0.520
avg_word_length,-0.030,0.137,1.000,-0.028,0.105,-0.022
dependency_count,0.997,0.978,-0.028,1.000,0.675,0.541
dependency_tree_depth,0.674,0.696,0.105,0.675,1.000,0.278
avg_dependency_distance,0.531,0.520,-0.022,0.541,0.278,1.000


In [ ]:
length_comparison = pd.concat(
    [
        df[["token_count"]],
        linguistic_features
    ],
    axis=1
)

print("Correlation with LaBSE token count:")

display(
    length_comparison.corr()["token_count"]
    .sort_values(ascending=False)
    .to_frame("correlation")
)

Correlation with LaBSE token count:


,correlation
token_count,1.000000
dependency_count,0.990415
word_count,0.984862
char_count,0.965680
dependency_tree_depth,0.648588
avg_dependency_distance,0.527068
avg_word_length,-0.024478


## Normalize structural measures

In [ ]:
complexity_features = linguistic_features.copy()

# Avoid division by zero
word_counts = complexity_features["word_count"].replace(0, np.nan)

complexity_features["dependency_density"] = (
    complexity_features["dependency_count"] / word_counts
)

complexity_features["dependency_density"] = (
    complexity_features["dependency_density"]
    .fillna(0.0)
)

complexity_features["dependency_distance_per_word"] = (
    complexity_features["avg_dependency_distance"]
    / word_counts
)

complexity_features["dependency_distance_per_word"] = (
    complexity_features["dependency_distance_per_word"]
    .replace([np.inf, -np.inf], np.nan)
    .fillna(0.0)
)

print("Updated feature set:")
print(complexity_features.columns.tolist())

display(complexity_features.head())

Updated feature set:
['word_count', 'char_count', 'avg_word_length', 'dependency_count', 'dependency_tree_depth', 'avg_dependency_distance', 'dependency_density', 'dependency_distance_per_word']


,word_count,char_count,avg_word_length,dependency_count,dependency_tree_depth,avg_dependency_distance,dependency_density,dependency_distance_per_word
0,19,105,4.578947,18,9,1.722222,0.947368,0.090643
1,42,236,4.642857,43,8,2.418605,1.023810,0.057586
2,19,101,4.368421,19,7,2.000000,1.000000,0.105263
3,16,101,5.375000,17,4,3.235294,1.062500,0.202206
4,32,192,5.031250,32,4,2.781250,1.000000,0.086914


In [ ]:
percentiles = [0.01, 0.05, 0.25, 0.50, 0.75, 0.95, 0.99]

distribution_table = complexity_features.quantile(percentiles).T

display(distribution_table)

,0.01,0.05,0.25,0.50,0.75,0.95,0.99
word_count,4.000000,7.000000,13.000000,18.000000,26.000000,45.000000,64.000000
char_count,19.000000,35.000000,68.000000,98.000000,140.000000,240.000000,340.480000
avg_word_length,3.064667,3.428571,3.923077,4.307692,4.750000,5.550000,6.285714
dependency_count,3.000000,6.000000,12.000000,18.000000,26.000000,45.000000,64.000000
dependency_tree_depth,1.000000,2.000000,4.000000,5.000000,7.000000,10.000000,13.000000
avg_dependency_distance,1.250000,1.571429,2.000000,2.400000,2.823529,3.600000,4.314079
dependency_density,0.750000,0.857143,0.931034,0.960000,1.000000,1.062500,1.115385
dependency_distance_per_word,0.043961,0.062091,0.097403,0.131313,0.175824,0.272189,0.380952


In [ ]:
for column in complexity_features.columns:

    print("\n" + "=" * 70)
    print(column)

    print("Minimum:", complexity_features[column].min())
    print("Maximum:", complexity_features[column].max())

    print("\nLargest values:")

    display(
        df.loc[
            complexity_features[column]
            .nlargest(5)
            .index,
            ["segment_id", "label_idx", "text"]
        ].assign(
            feature_value=complexity_features.loc[
                complexity_features[column].nlargest(5).index,
                column
            ].values
        )
    )


word_count
Minimum: 1
Maximum: 194

Largest values:


,segment_id,label_idx,text,feature_value
8814,83310,2,umm he wakes up twenty years later and uhh he ...,194
17309,dQ56b0bqmc8,2,at and all you need to do is print this form o...,194
8085,6UV6ktwbLoo,0,this is a quick video to show you how to do th...,188
7988,69707,3,cheesy it's just one of those really cheesy mo...,187
7894,6-0bcijTR8k,0,hi there today i have a very special announcem...,175



char_count
Minimum: 2
Maximum: 957

Largest values:


,segment_id,label_idx,text,feature_value
8085,6UV6ktwbLoo,0,this is a quick video to show you how to do th...,957
17309,dQ56b0bqmc8,2,at and all you need to do is print this form o...,951
8814,83310,2,umm he wakes up twenty years later and uhh he ...,947
7988,69707,3,cheesy it's just one of those really cheesy mo...,944
7894,6-0bcijTR8k,0,hi there today i have a very special announcem...,915



avg_word_length
Minimum: 1.6666666666666667
Maximum: 9.75

Largest values:


,segment_id,label_idx,text,feature_value
15916,ZSBoSJaD2tQ,6,dominates customer relationship management,9.750000
1258,126872,2,seriously,9.000000
889,111881,4,very disappointing,8.500000
10432,Ceu9AkIjuLY,1,increase accountability and effectiveness by c...,8.071429
3864,244261,15,fourteen fourteen,8.000000



dependency_count
Minimum: 0
Maximum: 200

Largest values:


,segment_id,label_idx,text,feature_value
17309,dQ56b0bqmc8,2,at and all you need to do is print this form o...,200
8814,83310,2,umm he wakes up twenty years later and uhh he ...,191
7988,69707,3,cheesy it's just one of those really cheesy mo...,189
8085,6UV6ktwbLoo,0,this is a quick video to show you how to do th...,188
7894,6-0bcijTR8k,0,hi there today i have a very special announcem...,184



dependency_tree_depth
Minimum: 0
Maximum: 25

Largest values:


,segment_id,label_idx,text,feature_value
18070,gXDRWhJbfFU,7,that includes the design for the cool cards th...,25
22841,yEtfjJOMMCs,10,it is very important if you need to take a res...,23
7316,4Vl6AeEkAg4,0,my name is carla santiago chair of the federat...,22
22768,y3r2kk8zvl0,7,on one hand they do allow you to think about h...,21
11707,I3SPlbTGuNc,6,exciting now that we are hard delivering on a ...,20



avg_dependency_distance
Minimum: 0.0
Maximum: 8.320512820512821

Largest values:


,segment_id,label_idx,text,feature_value
2258,206049,2,vince vaughn plays santa claus' nicolas claus'...,8.320513
18238,gvjQ1XfMA_g,2,mouse and more broadly than that whether it's ...,6.468085
9305,90172,8,uhh if you don't know what it's about uhh it's...,6.361111
5096,267466,14,it's all out there it's a little bit ridiculou...,6.250000
5539,272838,16,commit it's a like for profit prison where the...,6.089286



dependency_density
Minimum: 0.0
Maximum: 1.3333333333333333

Largest values:


,segment_id,label_idx,text,feature_value
3521,238060,2,uhh it's sickening it's a movie it's there's n...,1.333333
8340,74101,1,umm today's product that i'm gonna review is a...,1.300000
15238,Wh6sht7xwqQ,5,you'll working career you hadn't planned you'll,1.285714
3702,24196,13,he goes out we all know he's gonna fight he's ...,1.250000
4529,260011,0,what's up expo i've,1.250000



dependency_distance_per_word
Minimum: 0.0
Maximum: 0.75

Largest values:


,segment_id,label_idx,text,feature_value
4016,247382,0,hi i'm,0.750000
2502,21137,1,theaters it's called,0.666667
5342,270956,0,hi today i'm,0.666667
9333,91276,11,charisma doesn't work,0.666667
9526,96099,0,hi today i'm,0.666667


## Final linguistic complexity feature set

In [ ]:
FINAL_COMPLEXITY_FEATURES = [
    "word_count",
    "char_count",
    "avg_word_length",
    "avg_dependency_distance",
    "dependency_tree_depth",
    "dependency_density"
]

linguistic_complexity_df = complexity_features[
    FINAL_COMPLEXITY_FEATURES
].copy()

print("Final candidate linguistic-complexity features:")
for feature in FINAL_COMPLEXITY_FEATURES:
    print("-", feature)

print("\nShape:", linguistic_complexity_df.shape)

display(linguistic_complexity_df.head())

Final candidate linguistic-complexity features:
- word_count
- char_count
- avg_word_length
- avg_dependency_distance
- dependency_tree_depth
- dependency_density

Shape: (23253, 6)


,word_count,char_count,avg_word_length,avg_dependency_distance,dependency_tree_depth,dependency_density
0,19,105,4.578947,1.722222,9,0.947368
1,42,236,4.642857,2.418605,8,1.023810
2,19,101,4.368421,2.000000,7,1.000000
3,16,101,5.375000,3.235294,4,1.062500
4,32,192,5.031250,2.781250,4,1.000000


In [ ]:
assert linguistic_complexity_df.shape[0] == df.shape[0]

assert linguistic_complexity_df.isnull().sum().sum() == 0

assert np.isfinite(
    linguistic_complexity_df.to_numpy()
).all()

assert (
    linguistic_complexity_df.select_dtypes(include=np.number) >= 0
).all().all()

print("✓ Row count matches original dataset")
print("✓ No missing values")
print("✓ No infinite values")
print("✓ No negative values")

print("\nFinal statistics:")
display(linguistic_complexity_df.describe().T)

✓ Row count matches original dataset
✓ No missing values
✓ No infinite values
✓ No negative values

Final statistics:


,count,mean,std,min,25%,50%,75%,max
word_count,23253.0,21.122952,12.474859,1.000000,13.000000,18.000000,26.000000,194.000000
char_count,23253.0,112.300907,66.919655,2.000000,68.000000,98.000000,140.000000,957.000000
avg_word_length,23253.0,4.375601,0.660444,1.666667,3.923077,4.307692,4.750000,9.750000
avg_dependency_distance,23253.0,2.462472,0.638813,0.000000,2.000000,2.400000,2.823529,8.320513
dependency_tree_depth,23253.0,5.528491,2.452002,0.000000,4.000000,5.000000,7.000000,25.000000
dependency_density,23253.0,0.960386,0.074799,0.000000,0.931034,0.960000,1.000000,1.333333


In [ ]:
final_corr = linguistic_complexity_df.corr()

print("Final linguistic-complexity feature correlations:")

display(final_corr.round(3))

Final linguistic-complexity feature correlations:


,word_count,char_count,avg_word_length,avg_dependency_distance,dependency_tree_depth,dependency_density
word_count,1.000,0.981,-0.030,0.531,0.674,0.331
char_count,0.981,1.000,0.137,0.520,0.696,0.332
avg_word_length,-0.030,0.137,1.000,-0.022,0.105,0.046
avg_dependency_distance,0.531,0.520,-0.022,1.000,0.278,0.399
dependency_tree_depth,0.674,0.696,0.105,0.278,1.000,0.338
dependency_density,0.331,0.332,0.046,0.399,0.338,1.000


In [ ]:
linguistic_complexity_artifact = pd.concat(
    [
        df[
            [
                "segment_id",
                "label_idx",
                "start",
                "end",
                "text",
                "token_count"
            ]
        ].reset_index(drop=True),

        linguistic_complexity_df.reset_index(drop=True)
    ],
    axis=1
)

print("Artifact shape:", linguistic_complexity_artifact.shape)

print("\nColumns:")
print(linguistic_complexity_artifact.columns.tolist())

display(linguistic_complexity_artifact.head())

Artifact shape: (23253, 12)

Columns:
['segment_id', 'label_idx', 'start', 'end', 'text', 'token_count', 'word_count', 'char_count', 'avg_word_length', 'avg_dependency_distance', 'dependency_tree_depth', 'dependency_density']


,segment_id,label_idx,start,end,text,token_count,word_count,char_count,avg_word_length,avg_dependency_distance,dependency_tree_depth,dependency_density
0,--qXJuDtHPw,0,23.199,30.325,writer i see that a writer is somebody who has...,21,19,105,4.578947,1.722222,9,0.947368
1,-3g5yACwYnA,0,82.753,100.555,key is part of the people that we use to solve...,49,42,236,4.642857,2.418605,8,1.023810
2,-3g5yACwYnA,1,119.919,125.299,businesses that we do they've been able to fin...,23,19,101,4.368421,2.000000,7,1.000000
3,-3g5yACwYnA,2,4.840,14.052,operations key polymer brings a technical aspe...,22,16,101,5.375000,3.235294,4,1.062500
4,-3g5yACwYnA,3,13.211,27.521,internally we're a huge user of adhesives for ...,41,32,192,5.031250,2.781250,4,1.000000


In [ ]:
LINGUISTIC_COMPLEXITY_PATH = (
    "/content/drive/MyDrive/cmu_mosei/"
    "CMU_MOSEI_linguistic_complexity_features.pkl"
)

with open(LINGUISTIC_COMPLEXITY_PATH, "wb") as f:
    pickle.dump(
        linguistic_complexity_artifact,
        f
    )

print("Saved to:")
print(LINGUISTIC_COMPLEXITY_PATH)

print("\nFile exists:",
      os.path.exists(LINGUISTIC_COMPLEXITY_PATH))

Saved to:
/content/drive/MyDrive/cmu_mosei/CMU_MOSEI_linguistic_complexity_features.pkl

File exists: True


## Update: Have to drop the unwanted/redundant features

In [1]:
import pandas as pd
import numpy as np

LC_PATH = "/content/drive/MyDrive/cmu_mosei/CMU_MOSEI_linguistic_complexity_features.pkl"

df_lc = pd.read_pickle(LC_PATH)

print("Shape:", df_lc.shape)
print("\nColumns:")
print(df_lc.columns.tolist())

df_lc.head()

Shape: (23253, 12)

Columns:
['segment_id', 'label_idx', 'start', 'end', 'text', 'token_count', 'word_count', 'char_count', 'avg_word_length', 'avg_dependency_distance', 'dependency_tree_depth', 'dependency_density']


,segment_id,label_idx,start,end,text,token_count,word_count,char_count,avg_word_length,avg_dependency_distance,dependency_tree_depth,dependency_density
0,--qXJuDtHPw,0,23.199,30.325,writer i see that a writer is somebody who has...,21,19,105,4.578947,1.722222,9,0.947368
1,-3g5yACwYnA,0,82.753,100.555,key is part of the people that we use to solve...,49,42,236,4.642857,2.418605,8,1.023810
2,-3g5yACwYnA,1,119.919,125.299,businesses that we do they've been able to fin...,23,19,101,4.368421,2.000000,7,1.000000
3,-3g5yACwYnA,2,4.840,14.052,operations key polymer brings a technical aspe...,22,16,101,5.375000,3.235294,4,1.062500
4,-3g5yACwYnA,3,13.211,27.521,internally we're a huge user of adhesives for ...,41,32,192,5.031250,2.781250,4,1.000000


In [2]:
import pandas as pd
import numpy as np

LC_PATH = "/content/drive/MyDrive/cmu_mosei/CMU_MOSEI_linguistic_complexity_features.pkl"

df_lc = pd.read_pickle(LC_PATH)

print("Shape:", df_lc.shape)
print("\nColumns:")
print(df_lc.columns.tolist())

df_lc.head()

Shape: (23253, 12)

Columns:
['segment_id', 'label_idx', 'start', 'end', 'text', 'token_count', 'word_count', 'char_count', 'avg_word_length', 'avg_dependency_distance', 'dependency_tree_depth', 'dependency_density']


,segment_id,label_idx,start,end,text,token_count,word_count,char_count,avg_word_length,avg_dependency_distance,dependency_tree_depth,dependency_density
0,--qXJuDtHPw,0,23.199,30.325,writer i see that a writer is somebody who has...,21,19,105,4.578947,1.722222,9,0.947368
1,-3g5yACwYnA,0,82.753,100.555,key is part of the people that we use to solve...,49,42,236,4.642857,2.418605,8,1.023810
2,-3g5yACwYnA,1,119.919,125.299,businesses that we do they've been able to fin...,23,19,101,4.368421,2.000000,7,1.000000
3,-3g5yACwYnA,2,4.840,14.052,operations key polymer brings a technical aspe...,22,16,101,5.375000,3.235294,4,1.062500
4,-3g5yACwYnA,3,13.211,27.521,internally we're a huge user of adhesives for ...,41,32,192,5.031250,2.781250,4,1.000000


In [5]:
candidate_features = [
    "word_count",
    "char_count",
    "avg_word_length",
    "dependency_count",
    "dependency_tree_depth",
    "avg_dependency_distance",
    "dependency_density"
]

print("Candidate linguistic features:")
for feature in candidate_features:
    print(f"- {feature}")

Candidate linguistic features:
- word_count
- char_count
- avg_word_length
- dependency_count
- dependency_tree_depth
- avg_dependency_distance
- dependency_density


In [6]:
drop_features = [
    "char_count",
    "dependency_density"
]

df_lc_reduced = df_lc.drop(columns=drop_features)

print("Removed:")
for feature in drop_features:
    print(f"- {feature}")

print("\nRemaining columns:")
print(df_lc_reduced.columns.tolist())

print("\nShape:", df_lc_reduced.shape)

Removed:
- char_count
- dependency_density

Remaining columns:
['segment_id', 'label_idx', 'start', 'end', 'text', 'token_count', 'word_count', 'avg_word_length', 'avg_dependency_distance', 'dependency_tree_depth']

Shape: (23253, 10)


In [7]:
length_features = [
    "word_count"
]

surface_linguistic_features = [
    "avg_word_length"
]

syntactic_features = [
    "dependency_tree_depth",
    "avg_dependency_distance"
]

final_linguistic_features = (
    length_features
    + surface_linguistic_features
    + syntactic_features
)

print("Final linguistic-complexity candidates:")
for feature in final_linguistic_features:
    print(f"- {feature}")

Final linguistic-complexity candidates:
- word_count
- avg_word_length
- dependency_tree_depth
- avg_dependency_distance


In [8]:
df_lc_reduced[final_linguistic_features].describe().T

,count,mean,std,min,25%,50%,75%,max
word_count,23253.0,21.122952,12.474859,1.000000,13.000000,18.000000,26.000000,194.000000
avg_word_length,23253.0,4.375601,0.660444,1.666667,3.923077,4.307692,4.750000,9.750000
dependency_tree_depth,23253.0,5.528491,2.452002,0.000000,4.000000,5.000000,7.000000,25.000000
avg_dependency_distance,23253.0,2.462472,0.638813,0.000000,2.000000,2.400000,2.823529,8.320513


In [9]:
corr_matrix = df_lc_reduced[final_linguistic_features].corr()

print("Correlation matrix:")
display(corr_matrix.round(4))

Correlation matrix:


,word_count,avg_word_length,dependency_tree_depth,avg_dependency_distance
word_count,1.0000,-0.0301,0.6744,0.5308
avg_word_length,-0.0301,1.0000,0.1051,-0.0220
dependency_tree_depth,0.6744,0.1051,1.0000,0.2784
avg_dependency_distance,0.5308,-0.0220,0.2784,1.0000


In [10]:
from scipy.stats import pearsonr, spearmanr

features_to_control = [
    "dependency_tree_depth",
    "avg_dependency_distance"
]

control = df_lc_reduced["word_count"]

print("Raw correlation with word_count:\n")

for feature in features_to_control:
    pearson_r, pearson_p = pearsonr(
        df_lc_reduced[feature],
        control
    )

    spearman_r, spearman_p = spearmanr(
        df_lc_reduced[feature],
        control
    )

    print(f"{feature}")
    print(f"  Pearson  r = {pearson_r:.4f}, p = {pearson_p:.4e}")
    print(f"  Spearman r = {spearman_r:.4f}, p = {spearman_p:.4e}")
    print()

Raw correlation with word_count:

dependency_tree_depth
  Pearson  r = 0.6744, p = 0.0000e+00
  Spearman r = 0.7388, p = 0.0000e+00

avg_dependency_distance
  Pearson  r = 0.5308, p = 0.0000e+00
  Spearman r = 0.5647, p = 0.0000e+00



In [11]:
from scipy.stats import pearsonr
from sklearn.linear_model import LinearRegression

def partial_correlation(df, feature_x, feature_y, control_feature):
    data = df[[feature_x, feature_y, control_feature]].dropna()

    X_control = data[[control_feature]]

    # Residualize X
    model_x = LinearRegression()
    model_x.fit(X_control, data[feature_x])
    residual_x = data[feature_x] - model_x.predict(X_control)

    # Residualize Y
    model_y = LinearRegression()
    model_y.fit(X_control, data[feature_y])
    residual_y = data[feature_y] - model_y.predict(X_control)

    r, p = pearsonr(residual_x, residual_y)

    return r, p


for feature in features_to_control:
    r, p = partial_correlation(
        df_lc_reduced,
        feature,
        "avg_word_length",
        "word_count"
    )

    print(
        f"{feature} vs avg_word_length "
        f"(controlling word_count): "
        f"r = {r:.4f}, p = {p:.4e}"
    )

dependency_tree_depth vs avg_word_length (controlling word_count): r = 0.1699, p = 4.7299e-150
avg_dependency_distance vs avg_word_length (controlling word_count): r = -0.0071, p = 2.7936e-01


In [12]:
from sklearn.linear_model import LinearRegression

syntactic_features = [
    "dependency_tree_depth",
    "avg_dependency_distance"
]

for feature in syntactic_features:

    data = df_lc_reduced[
        ["word_count", feature]
    ].dropna()

    X = data[["word_count"]]
    y = data[feature]

    model = LinearRegression()
    model.fit(X, y)

    predicted = model.predict(X)
    residual = y - predicted

    df_lc_reduced.loc[data.index, f"{feature}_length_residual"] = residual

    r2 = model.score(X, y)

    print(f"{feature}")
    print(f"R² explained by word_count = {r2:.4f}")
    print(f"Residual standard deviation = {residual.std():.4f}")
    print()

dependency_tree_depth
R² explained by word_count = 0.4548
Residual standard deviation = 1.8106

avg_dependency_distance
R² explained by word_count = 0.2817
Residual standard deviation = 0.5414



In [13]:
residual_features = [
    "dependency_tree_depth_length_residual",
    "avg_dependency_distance_length_residual"
]

df_lc_reduced[residual_features].describe().T

,count,mean,std,min,25%,50%,75%,max
dependency_tree_depth_length_residual,23253.0,-5.280256e-16,1.810567,-19.443176,-1.186703,-0.247096,0.945846,11.125453
avg_dependency_distance_length_residual,23253.0,-2.695130e-16,0.541394,-3.971754,-0.381178,-0.044254,0.312595,4.339249


In [14]:
for feature in residual_features:

    valid = df_lc_reduced[
        ["word_count", feature]
    ].dropna()

    r, p = pearsonr(
        valid["word_count"],
        valid[feature]
    )

    print(f"{feature}")
    print(f"Pearson r with word_count = {r:.6f}")
    print(f"p-value = {p:.4e}")
    print()

dependency_tree_depth_length_residual
Pearson r with word_count = -0.000000
p-value = 1.0000e+00

avg_dependency_distance_length_residual
Pearson r with word_count = 0.000000
p-value = 1.0000e+00



In [15]:
bins = [0, 5, 10, 20, 40, 80, np.inf]
labels = [
    "1–5",
    "6–10",
    "11–20",
    "21–40",
    "41–80",
    "81+"
]

df_lc_reduced["length_bin"] = pd.cut(
    df_lc_reduced["word_count"],
    bins=bins,
    labels=labels,
    include_lowest=True
)

length_summary = (
    df_lc_reduced
    .groupby("length_bin", observed=True)[
        [
            "word_count",
            "dependency_tree_depth",
            "avg_dependency_distance"
        ]
    ]
    .agg(["count", "mean", "median"])
)

display(length_summary)

word_count                    dependency_tree_depth             \
                count        mean median                 count       mean   
length_bin                                                                  
1–5               538    3.966543    4.0                   538   1.644981   
6–10             2995    8.427045    9.0                  2995   3.114858   
11–20            9911   15.418928   15.0                  9911   4.827263   
21–40            8157   27.410568   26.0                  8157   6.797965   
41–80            1596   50.972431   48.0                  1596   9.043233   
81+                56  107.892857   96.0                    56  10.946429   

                  avg_dependency_distance                      
           median                   count      mean    median  
length_bin                                                     
1–5           2.0                     538  1.408736  1.500000  
6–10          3.0                    2995  1.964765  1.900000  
11–20         5.0                    9911  2.339181  2.294118  
21–40         6.0                    8157  2.720081  2.666667  
41–80         8.0                    1596  3.166463  3.046032  
81+          10.0                      56  3.437405  3.291861

In [18]:
variance_table = (
    df_lc_reduced[final_linguistic_features]
    .var()
    .sort_values()
)

print("Feature variances:")
display(variance_table)

Feature variances:


,0
avg_dependency_distance,0.408082
avg_word_length,0.436186
dependency_tree_depth,6.012316
word_count,155.622118


In [19]:
final_columns = [
    "segment_id",
    "label_idx",
    "start",
    "end",
    "text",
    "token_count",
    "word_count",
    "avg_word_length",
    "dependency_tree_depth",
    "avg_dependency_distance"
]

linguistic_complexity_df = df_lc_reduced[final_columns].copy()

print("Final linguistic complexity dataframe:")
print(linguistic_complexity_df.shape)

display(linguistic_complexity_df.head())

Final linguistic complexity dataframe:
(23253, 10)


,segment_id,label_idx,start,end,text,token_count,word_count,avg_word_length,dependency_tree_depth,avg_dependency_distance
0,--qXJuDtHPw,0,23.199,30.325,writer i see that a writer is somebody who has...,21,19,4.578947,9,1.722222
1,-3g5yACwYnA,0,82.753,100.555,key is part of the people that we use to solve...,49,42,4.642857,8,2.418605
2,-3g5yACwYnA,1,119.919,125.299,businesses that we do they've been able to fin...,23,19,4.368421,7,2.000000
3,-3g5yACwYnA,2,4.840,14.052,operations key polymer brings a technical aspe...,22,16,5.375000,4,3.235294
4,-3g5yACwYnA,3,13.211,27.521,internally we're a huge user of adhesives for ...,41,32,5.031250,4,2.781250


In [20]:
FINAL_LC_PATH = (
    "/content/drive/MyDrive/cmu_mosei/"
    "CMU_MOSEI_linguistic_complexity_features_v2.pkl"
)

linguistic_complexity_df.to_pickle(FINAL_LC_PATH)

print("Saved successfully:")
print(FINAL_LC_PATH)

print("\nFile exists:", __import__("os").path.exists(FINAL_LC_PATH))

Saved successfully:
/content/drive/MyDrive/cmu_mosei/CMU_MOSEI_linguistic_complexity_features_v2.pkl

File exists: True


In [21]:
loaded_final = pd.read_pickle(FINAL_LC_PATH)

print("Shape:", loaded_final.shape)

print("\nColumns:")
print(loaded_final.columns.tolist())

print("\nMissing values:")
print(loaded_final.isna().sum())

print("\nFeature statistics:")
display(
    loaded_final[
        [
            "word_count",
            "avg_word_length",
            "dependency_tree_depth",
            "avg_dependency_distance"
        ]
    ].describe().T
)

Shape: (23253, 10)

Columns:
['segment_id', 'label_idx', 'start', 'end', 'text', 'token_count', 'word_count', 'avg_word_length', 'dependency_tree_depth', 'avg_dependency_distance']

Missing values:
segment_id                 0
label_idx                  0
start                      0
end                        0
text                       0
token_count                0
word_count                 0
avg_word_length            0
dependency_tree_depth      0
avg_dependency_distance    0
dtype: int64

Feature statistics:


,count,mean,std,min,25%,50%,75%,max
word_count,23253.0,21.122952,12.474859,1.000000,13.000000,18.000000,26.000000,194.000000
avg_word_length,23253.0,4.375601,0.660444,1.666667,3.923077,4.307692,4.750000,9.750000
dependency_tree_depth,23253.0,5.528491,2.452002,0.000000,4.000000,5.000000,7.000000,25.000000
avg_dependency_distance,23253.0,2.462472,0.638813,0.000000,2.000000,2.400000,2.823529,8.320513


In [22]:
grouped_variation = (
    df_lc_reduced
    .groupby("length_bin", observed=True)
    [
        [
            "dependency_tree_depth",
            "avg_dependency_distance"
        ]
    ]
    .agg(["mean", "std", "median"])
)

display(grouped_variation)

dependency_tree_depth                  avg_dependency_distance  \
                            mean       std median                    mean   
length_bin                                                                  
1–5                     1.644981  0.792412    2.0                1.408736   
6–10                    3.114858  1.008065    3.0                1.964765   
11–20                   4.827263  1.478957    5.0                2.339181   
21–40                   6.797965  2.083326    6.0                2.720081   
41–80                   9.043233  2.894164    8.0                3.166463   
81+                    10.946429  4.354390   10.0                3.437405   

                                
                 std    median  
length_bin                      
1–5         0.514803  1.500000  
6–10        0.407085  1.900000  
11–20       0.488881  2.294118  
21–40       0.576573  2.666667  
41–80       0.731014  3.046032  
81+         0.879842  3.291861

In [23]:
comparison_df = df_lc_reduced[
    [
        "word_count",
        "dependency_tree_depth",
        "avg_dependency_distance",
        "dependency_tree_depth_length_residual",
        "avg_dependency_distance_length_residual"
    ]
].copy()

comparison_df.head(10)

,word_count,dependency_tree_depth,avg_dependency_distance,dependency_tree_depth_length_residual,avg_dependency_distance_length_residual
0,19,9,1.722222,3.752904,-0.682546
1,42,8,2.418605,-0.295723,-0.611325
2,19,7,2.000000,1.752904,-0.404768
3,16,4,3.235294,-0.849449,0.912069
4,32,4,2.781250,-2.970233,0.023130
5,36,10,2.371429,2.499571,-0.495416
6,24,9,2.260870,3.090159,-0.279803
7,32,12,1.645161,5.029767,-1.112959
8,16,5,2.562500,0.150551,0.239275
9,18,6,2.117647,0.885453,-0.259940


In [24]:
final_feature_corr = (
    df_lc_reduced[
        [
            "word_count",
            "avg_word_length",
            "dependency_tree_depth",
            "avg_dependency_distance"
        ]
    ]
    .corr()
)

display(final_feature_corr.round(4))

,word_count,avg_word_length,dependency_tree_depth,avg_dependency_distance
word_count,1.0000,-0.0301,0.6744,0.5308
avg_word_length,-0.0301,1.0000,0.1051,-0.0220
dependency_tree_depth,0.6744,0.1051,1.0000,0.2784
avg_dependency_distance,0.5308,-0.0220,0.2784,1.0000


In [25]:
final_columns = [
    "segment_id",
    "label_idx",
    "start",
    "end",
    "text",
    "token_count",
    "word_count",
    "avg_word_length",
    "dependency_tree_depth",
    "avg_dependency_distance"
]

linguistic_complexity_df = df_lc_reduced[final_columns].copy()

print("Final shape:", linguistic_complexity_df.shape)

display(linguistic_complexity_df.head())

Final shape: (23253, 10)


,segment_id,label_idx,start,end,text,token_count,word_count,avg_word_length,dependency_tree_depth,avg_dependency_distance
0,--qXJuDtHPw,0,23.199,30.325,writer i see that a writer is somebody who has...,21,19,4.578947,9,1.722222
1,-3g5yACwYnA,0,82.753,100.555,key is part of the people that we use to solve...,49,42,4.642857,8,2.418605
2,-3g5yACwYnA,1,119.919,125.299,businesses that we do they've been able to fin...,23,19,4.368421,7,2.000000
3,-3g5yACwYnA,2,4.840,14.052,operations key polymer brings a technical aspe...,22,16,5.375000,4,3.235294
4,-3g5yACwYnA,3,13.211,27.521,internally we're a huge user of adhesives for ...,41,32,5.031250,4,2.781250


In [26]:
feature_description = pd.DataFrame({
    "Feature": [
        "word_count",
        "avg_word_length",
        "dependency_tree_depth",
        "avg_dependency_distance"
    ],
    "Role": [
        "Utterance length / control variable",
        "Surface lexical-form characteristic",
        "Hierarchical syntactic complexity",
        "Linear syntactic organization / dependency complexity"
    ],
    "ABDA relevance": [
        "Controls for variation caused by segment duration/length",
        "Captures changes in lexical form independent of sentiment",
        "Captures hierarchical structural organization",
        "Captures syntactic dependency organization"
    ]
})

display(feature_description)

,Feature,Role,ABDA relevance
0,word_count,Utterance length / control variable,Controls for variation caused by segment durat...
1,avg_word_length,Surface lexical-form characteristic,Captures changes in lexical form independent o...
2,dependency_tree_depth,Hierarchical syntactic complexity,Captures hierarchical structural organization
3,avg_dependency_distance,Linear syntactic organization / dependency com...,Captures syntactic dependency organization


In [27]:
FINAL_LC_PATH = (
    "/content/drive/MyDrive/cmu_mosei/"
    "CMU_MOSEI_linguistic_complexity_features_v2.pkl"
)

linguistic_complexity_df.to_pickle(FINAL_LC_PATH)

print("Saved:", FINAL_LC_PATH)
print("Shape:", linguistic_complexity_df.shape)

Saved: /content/drive/MyDrive/cmu_mosei/CMU_MOSEI_linguistic_complexity_features_v2.pkl
Shape: (23253, 10)


In [28]:
loaded_final = pd.read_pickle(FINAL_LC_PATH)

print("Reloaded shape:", loaded_final.shape)

print("\nColumns:")
print(loaded_final.columns.tolist())

print("\nMissing values:")
print(loaded_final.isna().sum())

print("\nFinal linguistic features:")
display(
    loaded_final[
        [
            "word_count",
            "avg_word_length",
            "dependency_tree_depth",
            "avg_dependency_distance"
        ]
    ].describe().T
)

Reloaded shape: (23253, 10)

Columns:
['segment_id', 'label_idx', 'start', 'end', 'text', 'token_count', 'word_count', 'avg_word_length', 'dependency_tree_depth', 'avg_dependency_distance']

Missing values:
segment_id                 0
label_idx                  0
start                      0
end                        0
text                       0
token_count                0
word_count                 0
avg_word_length            0
dependency_tree_depth      0
avg_dependency_distance    0
dtype: int64

Final linguistic features:


,count,mean,std,min,25%,50%,75%,max
word_count,23253.0,21.122952,12.474859,1.000000,13.000000,18.000000,26.000000,194.000000
avg_word_length,23253.0,4.375601,0.660444,1.666667,3.923077,4.307692,4.750000,9.750000
dependency_tree_depth,23253.0,5.528491,2.452002,0.000000,4.000000,5.000000,7.000000,25.000000
avg_dependency_distance,23253.0,2.462472,0.638813,0.000000,2.000000,2.400000,2.823529,8.320513


Four features were retained for the linguistic-complexity feature group: word count, average word length, dependency tree depth, and average dependency distance. Word count is retained primarily as an utterance-length/control variable. Dependency count and character count were excluded because they exhibited very high correlation with utterance length (r ≈ 0.99 and r ≈ 0.97, respectively), indicating substantial redundancy. Dependency density was also excluded because it is derived from dependency count and therefore provides little independent information. Dependency tree depth and average dependency distance were retained because, although both increase with utterance length, they retain substantial residual variation after accounting for length (45.5% and 71.8% of their variance, respectively, remains unexplained by word count).